### PR TITLE
display latitude longitute in URL

### PR DESCRIPTION
**Pull request by amartyahatua**
Displaying the latitude and longitude (to three decimal points) in the URL.

### CODE REVIEW

**Comment by amartyahatua** [comment]
If latitude and longitude of URL are not in the proper range, reset them to default values.

File: src/components/GlobeOptimized.jsx [hunk 1]
```diff
@@ -334,7 +334,10 @@ export default function GlobeOptimized(props) {
 
   function dragged(event) {
     const { dx, dy } = event;
-
+    let newParams = new URLSearchParams(window.location.search);
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx:388]: We need `debounce` here to limit the # of times we change browser's history. Otherwise the application will not be usable if we change the url per drag event.

File: src/components/GlobeOptimized.jsx [hunk 2]
```diff
@@ -395,20 +398,44 @@ export default function GlobeOptimized(props) {
       properties.forEach((property) => {
         switch (property) {
           case 'lat':
+            console.log('Rotation -->');
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: Can we remove those console logs?

File: src/components/GlobeOptimized.jsx [hunk 3]
```diff
@@ -395,20 +398,44 @@ export default function GlobeOptimized(props) {
...
+            console.log(rotation);
             rotation[1] +=
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: to simplify the logic here, we should verify the value first before assigning to `rotation[1]`, same applies to line 407 - 411

File: src/components/GlobeOptimized.jsx [hunk 4]
```diff
@@ -395,20 +398,44 @@ export default function GlobeOptimized(props) {
...
               params.has(property) && !isNaN(params.get(property))
                 ? Number(params.get(property))
                 : 0;
+            //if initial longitude and latitude not in the range rest it
+            if(rotation[1] <-360 || rotation[1] > 360){
+              rotation[1] = 0;
+              rotation[0] = 180;
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: Also why are we change `rotation[0]` when we verify the value for `rotation[1]` ?

File: src/components/GlobeOptimized.jsx [hunk 5]
```diff
@@ -395,20 +398,44 @@ export default function GlobeOptimized(props) {
...
+            }
             break;
           default:
+            
         }
       });
+
+      if((-360 > rotation[0] &&  rotation[0] > 360) ||(-360 > rotation[1] &&  rotation[1] > 360))
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: do you mind explaining this part?

File: src/components/GlobeOptimized.jsx [hunk 6]
```diff
@@ -395,20 +398,44 @@ export default function GlobeOptimized(props) {
...
         switch (property) {
           case 'lat':
+            console.log('Rotation -->');
+            console.log(rotation);
             rotation[1] +=
               params.has(property) && !isNaN(params.get(property))
                 ? Number(params.get(property))
                 : 0;
+            //if initial longitude and latitude not in the range rest it
+            if(rotation[1] <-360 || rotation[1] > 360){
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: We need to make sure the number here reflects the value people normally use when presenting the geo location. For example, we use 37 degree N (+37), 122 degree W (-122) to locate San Francisco city. So we need to make sure those values can be parsed into the proper rotation value.

File: src/components/GlobeOptimized.jsx (hunk 4 above)
Comment (amartyahatua) [reply src/components/GlobeOptimized.jsx]: If any of these values are out of the bound we are setting the globe to its initial position.

File: src/components/GlobeOptimized.jsx [hunk 7]
```diff
@@ -493,7 +509,7 @@ export default function GlobeOptimized(props) {
     // segment star payload
     (stars || []).forEach((star) => {
       const { location, mag } = star;
-      const adjustedMag = parseFloat(mag) * 10;
+      const adjustedMag = parseFloat(mag) * 15;
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: Did we confirm with @sidgan that we want to increase the magnitude of the stars?

File: src/components/GlobeOptimized.jsx [hunk 8]
```diff
@@ -436,8 +451,9 @@ export default function GlobeOptimized(props) {
 
     canvas.call(
       // drag().on('start', dragstarted).on('drag', dragged).on('end', dragended)
-      drag().on('drag', dragged)
+      drag().on('drag', debounce(dragged, 5))
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: We don't want to debounce the callback, otherwise it's going to cause some issue as the rotation parameter is not update to date. we should instead debounce the port where the query parameter is updated.

File: src/components/GlobeOptimized.jsx [hunk 9]
```diff
@@ -395,18 +402,26 @@ export default function GlobeOptimized(props) {
...
+              else{
+                rotation[1] = 0
+              }
             break;
           case 'long':
-            rotation[0] +=
-              params.has(property) && !isNaN(params.get(property))
-                ? Number(params.get(property))
-                : 180;
+            if(params.has(property) && !isNaN(params.get(property)) && ((-360 <= Number(params.get(property))) && (Number(params.get(property)) <= 360))){
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: I think this is still the part that concerns me as `-360` to `+360` is not the right range we wanna accept from the user.  For longitude, `-180` to `+180` would be the right range and for latitude, `-90` to `+90`. Please use [this comment](https://github.com/sidgan/nasaweb/pull/35#discussion_r753941715) as an example.

**Comment by sidgan** [comment]
Hi @chichengren can you review this so that its ready to be deployed?

File: src/components/GlobeOptimized.jsx [hunk 10]
```diff
@@ -325,6 +327,22 @@ export default function GlobeOptimized(props) {
     render();
   }
 
+
+  function map_latitude(latitude_int)
+  {
+    latitude_int = latitude_int < -90 && latitude_int >= -180 ? latitude_int + 90 : latitude_int < -180 && latitude_int >= -270 ? latitude_int + 180 : latitude_int < -270 && latitude_int >= -360 ? (latitude_int + 270)*(-1) : latitude_int
```
Comment (chichengren) [inline src/components/GlobeOptimized.jsx]: The tertiary operator here does make it really messy and hard to follow, do you mind breaking it into several if/else cases?

**Review by chichengren — APPROVED** [review]
